### PR TITLE
MAINT: stats._moment: warn when catastrophic cancellation occurs

### LINKED ISSUE #15554: BUG: stats.skew: unreliable when all values in axis slice are identical
### Describe your issue.

computing the skewness of a numpy ndarray will produce two result with following method for a column in ndarray (x):
1. skew(x[:, i])
2. skew(x)[i]

the issues happens when this column has equal float value

### Reproducing Code Example

```python
import numpy as np
from scipy.stats import skew

a = np.random.randn(100, 10)
a[:, 0] = 1.01

print(f'result in matrix value {skew(a)[0]}')   # -1.0
print(f'result in array value {skew(a[:, 0])}') # 0.0
```

### Error message

```shell
No error message
```

### SciPy/NumPy/Python version information

1.7.3 1.22.0 sys.version_info(major=3, minor=8, micro=5, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-15554
gh-13254
gh-13245
gh-11086
gh-10896

#### What does this implement/fix?
We get reports every once in a while of `skew` and/or `kurtosis` being unreliable with degenerate input. This PR would report a warning rather than trying to produce a correct result. Alternatively, we could replace the values that suffer from severe precision loss with `nan`. Re-reading gh-13254, I still don't think `0` is necessarily correct.

### CODE REVIEW

File: scipy/stats/_stats_py.py [hunk 1]
```diff
@@ -1013,6 +1013,14 @@ def _moment(a, moment, axis, *, mean=None):
         # Starting point for exponentiation by squares
         mean = a.mean(axis, keepdims=True) if mean is None else mean
         a_zero_mean = a - mean
+
+        eps = np.finfo(a_zero_mean.dtype).resolution * 10
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:1017]: 10? 100? 1000? How much precision loss is too much?

File: scipy/stats/_stats_py.py [hunk 2]
```diff
@@ -1013,6 +1013,14 @@ def _moment(a, moment, axis, *, mean=None):
...
+        rel_diff = np.max(np.abs(a_zero_mean), axis=axis, keepdims=True)/mean
+        if np.any((0 < rel_diff) & (rel_diff < eps)):
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:1019]: We could remove the `(0 < rel_diff)` but there are some tests with scalar input that start failing, and I wasn't sure whether we wanted to just warn or also return nan.

**Comment by mdhaber** [comment]
Should I break some of this out into a function and use it to address gh-14418, too? I suspect that is not the last place we'll see catastrophic cancellation, either.

File: scipy/stats/_stats_py.py (hunk 1 above)
Comment (tupui) [reply scipy/stats/_stats_py.py:1017]: I suppose we can start conservative like so and update it if it's too annoying for users.

Comment (tupui) [reply scipy/stats/_stats_py.py:1017]: Mmm but 1 digit of loss is better than 2 no? That's what I interpreted as conservative

Comment (mdhaber) [reply scipy/stats/_stats_py.py:1017]: Yes, but _all but one_ digit loss is worse than _all but two_ digits lost (15 digits lost vs 14 digits lost). A factor of 10 here corresponds with 15 digits lost.

Comment (tupui) [reply scipy/stats/_stats_py.py:1017]: Yes ok we are talking about the same. Conservative for me meant more "be strict" as having a high number would not bother anyone and show any warnings.

Comment (mdhaber) [reply scipy/stats/_stats_py.py:1017]: Ok. By "like so", I thought you meant "like it is now,  which very rarely emits the warning.
 In any case, if we were to be conservative by using, say, 10^6 here, the warning would be emitted whenever ~10 or more digits are lost, which is more frequent than as it is now: with 10^1 here, we emit only when 15 digits are lost. Where do you think the sweet spot is?

Comment (tupui) [reply scipy/stats/_stats_py.py:1017]: Personally I never need that much precision (because I have so much uncertainty elsewhere that precision here would be artificial) so I am not sure. Maybe 10 digits is enough.

Comment (mdhaber) [inline scipy/stats/_stats_py.py:1017]:
```suggestion
        eps = np.finfo(a_zero_mean.dtype).resolution * 1e6
```

Comment (mdhaber) [reply scipy/stats/_stats_py.py:1017]: No warnings locally; let's try it.

Comment (mdhaber) [reply scipy/stats/_stats_py.py]: Thanks @tupui I changed it to 1e6 so that we only warn if ~10 or more digits are lost.

**Comment by SeanChen1992** [comment]
您的邮件我已收到，我会加班加点的看完您珍贵的来信，尽快给您回复的！谢谢啦！陈鑫

File: scipy/stats/_stats_py.py [hunk 3]
```diff
@@ -1013,6 +1013,16 @@ def _moment(a, moment, axis, *, mean=None):
...
         mean = a.mean(axis, keepdims=True) if mean is None else mean
         a_zero_mean = a - mean
+
+        eps = np.finfo(a_zero_mean.dtype).resolution * 1e6
+        rel_diff = np.max(np.abs(a_zero_mean), axis=axis, keepdims=True)/mean
+        with np.errstate(invalid='ignore'):
+            precision_loss = np.any((0 < rel_diff) & (rel_diff < eps))
+        if precision_loss:
+            message = ("Precision loss occurred in moment calculation due to "
+                       "catastrophc cancellation. Results may be unreliable.")
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:1023]:
```suggestion
                       "catastrophic cancellation. Results may be unreliable.")
```

File: scipy/stats/_stats_py.py [hunk 4]
```diff
@@ -1013,6 +1013,16 @@ def _moment(a, moment, axis, *, mean=None):
         # Starting point for exponentiation by squares
         mean = a.mean(axis, keepdims=True) if mean is None else mean
         a_zero_mean = a - mean
+
+        eps = np.finfo(a_zero_mean.dtype).resolution * 1e6
+        rel_diff = np.max(np.abs(a_zero_mean), axis=axis, keepdims=True)/mean
+        with np.errstate(invalid='ignore'):
+            precision_loss = np.any((0 < rel_diff) & (rel_diff < eps))
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:1020]:
```suggestion
            precision_loss = np.any(rel_diff < eps)
```
We probably _do_ want this warning to show when all the data is exactly numerically equal.

File: scipy/stats/_stats_py.py [hunk 5]
```diff
@@ -1013,6 +1013,16 @@ def _moment(a, moment, axis, *, mean=None):
         # Starting point for exponentiation by squares
         mean = a.mean(axis, keepdims=True) if mean is None else mean
         a_zero_mean = a - mean
+
+        eps = np.finfo(a_zero_mean.dtype).resolution * 1e6
+        rel_diff = np.max(np.abs(a_zero_mean), axis=axis, keepdims=True)/mean
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:1018]: @thomasjpfan thanks for your help today. This issue is related to that one (gh-10821) in that it also involves catastrophic cancellation when computing a moment from data.

In gh-10821, you looked for the variance being zero (`< np.finfo(dtype).tiny`) and did something to compensate. When this happened, you didn't really need to calculate the variance very accurately because this signaled that the optimization need not consider that value of `lmbda`.

This is general n-th order moment computation code, so there's not much we can do to compensate when precision loss occurs. So here, I'm looking for the relative difference between data and the mean being less than some tolerance, and I just want to warn the user that some sort of precision loss has occurred. I could probably be more specific about how much and exactly where it has occurred, but that can be left to a follow-up.

Do you have any thoughts on this PR?

Comment (thomasjpfan) [reply scipy/stats/_stats_py.py:1018]: It's a trade off to me. Adding a precision loss check adds more computational, but it gives warnings when something looks wrong. My concern is the heuristic for checking the precision loss and making sure that does not warn for valid input. For example, how was `1e6` decided when computing `eps`?

As for the PR, if you want the bare mimumum change, one can directly detect the near constant case in `skew` and raise a specific warning for that.

Comment (mdhaber) [reply scipy/stats/_stats_py.py:1018]: Thanks @thomasjpfan!  We already have an exact equality check in `skew` and `kurtosis`, but afterward we got a another report for which that check was insufficient. Also, we've gotten the same sort of issue in a few contexts, so I thought I'd treat more broadly. 
`1e6` was arbitrary, but it corresponds with a loss of approximately 10 decimal digits of precision for `float64`. This may be too noisy, so I'll reduce it so that this only warns when there is almost complete loss of precision.
Thanks!

File: scipy/stats/_stats_py.py (hunk 3 above)
Comment (tupui) [inline scipy/stats/_stats_py.py:1023]: I think the message is just missing an information for the user. Is this actionable or not for them. If you get the warning, then what? I would add a line in that sense.

Comment (tupui) [reply scipy/stats/_stats_py.py:1023]: Maybe we can recommend to play with dtype or say in which condition this usually happen?

Comment (tupui) [reply scipy/stats/_stats_py.py:1023]: Better yes 👍

Comment (mdhaber) [reply scipy/stats/_stats_py.py]: Committed. Thanks! That will be helpful.

File: scipy/stats/_stats_py.py (hunk 4 above)
Comment (tirthasheshpatel) [reply scipy/stats/_stats_py.py:1020]: > We probably do want this warning to show when all the data is exactly numerically equal.

Yes, agreed.

> I don't think it was correct in [gh-13254](https://github.com/scipy/scipy/pull/13254) to set the `skew`/`kurtosis` to zero when all data were equal; it should be `nan`, and there should be a warning.

I agree: 0 is not the correct skew/kurtosis for near-constant inputs.

File: scipy/stats/_stats_py.py [hunk 6]
```diff
@@ -1013,6 +1013,17 @@ def _moment(a, moment, axis, *, mean=None):
         # Starting point for exponentiation by squares
         mean = a.mean(axis, keepdims=True) if mean is None else mean
         a_zero_mean = a - mean
+
+        eps = np.finfo(a_zero_mean.dtype).resolution * 10
+        rel_diff = np.max(np.abs(a_zero_mean), axis=axis, keepdims=True)/mean
```
Comment (tirthasheshpatel) [inline scipy/stats/_stats_py.py:1018]: Shouldn't this be:

```suggestion
        rel_diff = np.max(np.abs(a_zero_mean), axis=axis, keepdims=True) / np.abs(mean)
```

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel. Fingers crossed that I caught all the new warnings.

**Review by tirthasheshpatel — APPROVED** [review]
LGTM. Thanks @mdhaber! @tupui @thomasjpfan can one of you please take a look?

**Review by tupui — APPROVED** [review]
LGTM as well. Thanks everyone!

**Comment by mdhaber** [comment]
Great. @tirthasheshpatel would you like to submit the follow-up that changes what was done in gh-13254 so that `skew` and `kurtosis` return `nan` when input is exactly constant?

And I'll use this idea to fix some of the related issues?

**Comment by tirthasheshpatel** [comment]
> would you like to submit the follow-up that changes what was done in [gh-13254](https://github.com/scipy/scipy/pull/13254) so that `skew` and `kurtosis` return `nan` when input is exactly constant?

Sure!

> And I'll use this idea to fix some of the related issues?

:+1:
